Parallelize stereo fusion; needs pre-loading of entire workspace

anmatako opened the pull request:
Stereo fusion can now be done in parallel as long as the entire workspace gets pre-loaded. A new `NoCacheWorkspace` class can do that when initializing fusion. Parallel execution is the new default so to use the existing functionality with cache and sequential fusion `stereo_fusion` must be used with `--StereoFusion.use_cache 1`.

Also added options to:
- Use image masks that excludes pixels from the depth maps. Work similarly to masks in `ImageReaderOptions`
- Restrict the fused point cloud within a specified bounding box described from its two 3D corner points. The bounding box is in plan text format with space separated numeric values for the corner points, e.g. `<x1> <y1> <z1> <x2> <y2> <z2>.

anmatako commented:
This will not build for now since it has dependencies on #1129

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -292,31 +326,54 @@ void StereoFusion::Run() {
+  }
+}
+
+void StereoFusion::Fuse(int tid) {
```
tid->thread_id

anmatako replied:
Fixed

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -383,25 +440,35 @@ void StereoFusion::Fuse() {
     // Set the current pixel as visited.
     fused_pixel_mask.Set(row, col, true);
```
You will end up with race conditions here, because you simultaneously read/write to the masks from multiple threads.

anmatako replied:
Yes this is true, and ideally this should have been an atomic update, but unfortunately the OpenMP standard (at least on Windows) does not allow atomic operations with function calls, only simple updates like `x = val` or `x += val`, etc. Wrapping this in a critical section kills the performance since all threads get blocked instead of just the ones that may conflict so this was also a no-go from a performance standpoint.

I've let that race condition exist since all threads are only attempting to set the same value (i.e. make the mask value `true`) so they're not adversely impacting the result of fusion. Worst case, by my understanding, is that we reconsider a pixel that should have already been masked out as visited. This causes some potential duplication of work, but from a performance standpoint its practically negligible, and definitely much preferred to using a critical section.

If there is a better way to go about it I can update accordingly. If it so happens that atomic works better in Linux/Mac I can add the `#pragma omp atomic` with a specific guard to skip it on windows. Otherwise I would have to change the data type of the mask from `Mat<bool>` to `std::vector` or something similar so I can directly access the data without a need for a function call.

anmatako replied:
I think I managed to make this work with `#pragma omp atomic` by accessing the underlying pointer of `Mat<bool>`. Of course I'll have to performance test it on my end to ensure there is no adverse performance impact.

anmatako replied:
What I tried won't work either. Apparently `std::vector<bool>` does not have a `.data()` member, unlike other types of vectors. I managed to make it work by converting the mask to `Mat<int>` instead of `Mat<bool>` and doing this:
```
const int pixel_index = depth_map_sizes_.at(image_idx).first * row + col;
#pragma omp atomic
fused_pixel_mask.GetPtr()[pixel_index]++;
```

Note that I'm using increment and not assignment here since Windows only support OpenMP 2.0 and atomic only supports `+=`, `-=`, `++`, and `--`.

If you think that's a valid way of handling this then I'll go ahead and update the PR. Thankfully making this atomic does not impact performance for parallel execution, at least in the few datasets I tested.

ahojnnes replied:
I think your suggested fix doesn't resolve the actual problem causing potentially many duplicated fused points. I think this effect is exacerbated by the fact that you start spawning threads for each pixel row, i.e. nearby pixels will be processed by different threads, which are very likely to be fused into the same point.

I think we could minimize this effect by using some striding between different threads. For example:
```
const int kRowStride = max(1, height / num_threads);
#pragma omp parallel for schedule(dynamic) num_threads(num_threads)
for (int row_start = 0; row_start < height; row_start += kRowStride) {
    const int row_end = min(height, row_start + kRowStride);
    for (int row = row_start; row < row_end; ++row) {
        for (int col = 0; col < width; ++col) {
            ...
        }
    }
}
```
In this case, elements in the fusion queue of different threads are further apart and are less likely to race with each other.

I was not so much concerned about the atomic update, which will obviously be a problem for `bool`, which is typically implemented using bitshifting. We could switch to `Mat<char>` and just use `mask.Set(row, col, 1)`.

anmatako replied:
Yes, you are right, the current setup with dynamic scheduling does suffer from potential duplications. I think an easier approach could be to configure it with a batch size of 10 or more so that different threads start further apart; this avoids the need for manual management of offsets. I can experiment with a few different batch sizes and depth map sizes and see if there's some value that works well enough.

And yes doing `Mat<char>` for mask should work even without a specific guard for the atomic update, which will keep things simpler.

These two I can do and quickly update, but changing the format of fusion to work with the existing thread pool will take a while as it's not clear to me what kind of setup it will need to process the problems, and I don't want to make the approach too verbose. But, I'll definitely attempt this at some point.

anmatako replied:
So I just tested the current setup (chunk size of 1) against the single threaded run and multi-threaded runs with chunk sizes of 10, 20, and 50 (that should be the same as setting a row stride).

Interestingly, the processing time goes slightly up when increasing the chunk size, and the reconstructed points are slightly less than the single threaded point clouds, with the exception of chunk size of 1. Below is a summary of the findings.

- Reference single threaded run: Runtime 11.5 min; points: 7,570,936

The comparison results show the relative number of points and the % change:
1. Chunk size 1 (current schedule): Runtime: 1.45 min; points: +20,568 (+0.27%)
2. Chunk size 10: Runtime: 1.50 min; points: -13353 (-0.18%)
3. Chunk size 20: Runtime: 1.66 min; points: -6130 (-0.08%)
4. Chunk size 50: Runtime: 2.09 min; points: -2552 (-0.03%)

Those differences are really small, thankfully, and in light of that I think it would make sense to configure the scheduling with a chunk size of 10.

ahojnnes replied:
Chunk size of 10 seems like a good trade off. Thanks!

ahojnnes replied:
No need to do the thread pool - was just a suggestion.

anmatako replied:
Seeing how colmap does not natively use OpenMP in favor of the ThreadPool, I'm afraid that a solution with OpenMP would pollute the code with quite a few guards checking if OpenMP is available.

I think I have a clear idea now on how to go about this with the ThreadPool, so I'll give it a try and update the PR. Should be cleaner overall with that approach.

ahojnnes replied:
Looks great, thanks.

ahojnnes commented on src/exe/colmap.cc:
```diff
@@ -267,6 +268,7 @@ int RunStereoFuser(int argc, char** argv) {
   std::string pmvs_option_name = "option-all";
   std::string output_type = "PLY";
   std::string output_path;
+  std::string bbox_path;
```
The bounding box code here seems unrelated to the rest of the PR?

ahojnnes replied:
Ignore my comment. I just saw that "bounds" is used further below.

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -246,27 +277,30 @@ void StereoFusion::Run() {
+        data.row = row;
+        data.col = col;
+
+        const int tid = omp_get_thread_num();
```
thread_id

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -381,27 +438,40 @@ void StereoFusion::Fuse() {
+    mask_ptr[row * width + col] = true;
+
+    // Pixels out of bounds are filtered
+    if (xyz(0) < options_.bounds.first(0) ||
```
bounds -> bbox?

anmatako replied:
Will change

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -246,27 +277,30 @@ void StereoFusion::Run() {
-    for (data.row = 0; data.row < height; ++data.row) {
-      for (data.col = 0; data.col < width; ++data.col) {
-        if (fused_pixel_mask.Get(data.row, data.col)) {
+#pragma omp parallel for schedule(dynamic) num_threads(num_threads)
```
Any OpenMP code needs to be wrapped in `#ifdef OPENMP_ENABLED`, because it's an optional dependency in Python. Alternatively and preferably, we could use the ThreadPool in `util/threading.h`. We could implement a

anmatako replied:
Got it; I'll wrap it as such, and use your suggestion for an attempt to use the existing ThreadPool.

anmatako replied:
Quick question on wrapping with `#ifdef OPENMP_ENABLED`. I thought that the `#pragma`s are ignored in cases where they cannot be resolved (i.e. OpenMP missing and finding a `#pragma omp` directive). Is this not the case with the colmap setup?

ahojnnes replied:
It will certainly trigger an unknown pragma warning. If you then also treat warnings as errors, compilation will fail.

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -151,12 +157,27 @@ void StereoFusion::Run() {
+  } else {
+    workspace_.reset(new NoCacheWorkspace(workspace_options));
+    workspace_->Load(image_names);
+    num_threads = omp_get_max_threads();
```
It would be nice to make the number of threads configurable in the options. It could be initialized to -1 and you use `GetEffectiveNumThreads` to determine the number of threads.

anmatako replied:
Yes, that's a good suggestion and an easy change to make :)

anmatako commented:
Replaced OpenMP with ThreadPool for both fusion and workspace. Also removed all the members from Fusion that were only really used internally in the `Fuse` function. Now there's much less of the ugliness with the `thread_id` indexing. Performance took a minor hit, but overall it's not significant.

ahojnnes commented on src/exe/colmap.cc:
```diff
@@ -297,6 +300,20 @@ int RunStereoFuser(int argc, char** argv) {
+      auto& max_bound = options.stereo_fusion->bounding_box.second;
+      file >> min_bound(0) >> min_bound(1) >> min_bound(2);
+      file >> max_bound(0) >> max_bound(1) >> max_bound(2);
+      file.close();
```
.close() is not necessary as the destructor will do the job.

ahojnnes commented on src/mvs/workspace.h:
```diff
@@ -42,12 +42,17 @@
 
+class ConfidenceMap;
+
 class Workspace {
```
Maybe it's better to called to `NoCache` version `Workspace` and the cached one as `CachedWorkspace`.

anmatako replied:
OK sounds good, I'll flip the inheritance structure around.

ahojnnes commented on src/mvs/workspace.cc:
```diff
@@ -234,5 +235,88 @@ void ImportPMVSWorkspace(const Workspace& workspace,
+  confidence_maps_.resize(num_images);
+  normal_maps_.resize(num_images);
+
+  auto LoadWorkspaceData = [&, this](int image_idx) {
```
const int image_idx (in COLMAP we use const even for basic types)

ahojnnes commented on src/mvs/workspace.cc:
```diff
@@ -234,5 +235,88 @@ void ImportPMVSWorkspace(const Workspace& workspace,
+    }
+  };
+
+  int num_threads = GetEffectiveNumThreads(options_.num_threads);
```
const

ahojnnes commented on src/mvs/workspace.cc:
```diff
@@ -234,5 +235,88 @@ void ImportPMVSWorkspace(const Workspace& workspace,
+  std::cout << StringPrintf("Loading workspace data with %d threads...",
+                            num_threads)
+            << std::endl;
+  for (int i = 0; i < image_names.size(); ++i) {
```
size_t i

ahojnnes commented on src/mvs/fusion.h:
```diff
@@ -120,41 +137,36 @@ class StereoFusion : public Thread {
+    int image_idx;
+    int row;
+    int col;
+    int traversal_depth;
```
Can you restore the previous default values here? It's generally good practice to keep these to avoid uninitialized memory, I think.

anmatako replied:
Sounds good; I only removed them because now we never create a default `FusionData` struct and we only use it with the constructor through `emplace_back`

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -418,57 +500,52 @@ void StereoFusion::Fuse() {
           P_.at(next_image_idx) * xyz.homogeneous();
-      next_data.col = static_cast<int>(std::round(next_proj(0) / next_proj(2)));
-      next_data.row = static_cast<int>(std::round(next_proj(1) / next_proj(2)));
+      int next_col = static_cast<int>(std::round(next_proj(0) / next_proj(2)));
```
const and also next_row

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -292,35 +341,61 @@ void StereoFusion::Run() {
+  }
+}
+
+void StereoFusion::Fuse(int thread_id, int image_idx, int row, int col) {
```
please make the parameters const

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -235,7 +256,28 @@ void StereoFusion::Run() {
+            << std::endl;
+  ThreadPool thread_pool(num_threads);
+
+  const int kRowStride = 10;
```
It would be great to add a short explanation, why we use a stride here.

ahojnnes commented on src/mvs/fusion.cc:
```diff
@@ -246,40 +288,47 @@ void StereoFusion::Run() {
-  fused_points_visibility_.shrink_to_fit();
+  fused_points_.reserve(total_fused_points);
+  fused_points_visibility_.reserve(total_fused_points);
+  for (size_t task_id = 0; task_id < task_fused_points_.size(); ++task_id) {
```
task_id -> thread_id for consistency

ahojnnes commented:
@anmatako You will have to merge with master. note that I modularized exe/colmap.cc into individual source files, since it grew way beyond what we should put into a single source file.

anmatako commented:
> @anmatako You will have to merge with master. note that I modularized exe/colmap.cc into individual source files, since it grew way beyond what we should put into a single source file.

yes now that we are getting closer with these changes I'll do a merge from master and also resolve some other issues with confidence maps and calculated normals that were introduced from the PatchMatchNet PR.

I'll do a good cleanup and make sure this builds standalone.

ahojnnes commented on src/mvs/workspace.h:
```diff
@@ -63,34 +67,78 @@ class Workspace {
-  const Bitmap& GetBitmap(const int image_idx);
-  const DepthMap& GetDepthMap(const int image_idx);
-  const NormalMap& GetNormalMap(const int image_idx);
+  virtual inline const Bitmap& GetBitmap(const int image_idx) {
```
I suggest moving these method definitions to the .cc file. Since it's a virtual method, it won't be able to inline these anyway.

ahojnnes commented:
@anmatako Thanks, the PR looks great now. I trust you checked correctness that these changes lead to equivalent (+- small changes) fusion results as before? I left two minor comments, then I am happy to merge. Thanks!

anmatako commented:
> @anmatako Thanks, the PR looks great now. I trust you checked correctness that these changes lead to equivalent (+- small changes) fusion results as before? I left two minor comments, then I am happy to merge. Thanks!

Thanks! I've tested quite a lot in our pipelines and the differences are really minor compared to single threaded fusion; less than 0.1% in the number fused points and visually nothing stood out as an obvious outlier.

ahojnnes commented:
Thanks very much. This is a great improvement!
